### PR TITLE
Updated vbf filter v2

### PR DESCRIPTION
**Pull request by gmandorl**
VBFGenJetFilter.cc has been extended.
Possible requirements on the invariant mass of the two pt-leading genjets.
The considered genjets can also be asked to have a minimum DR from the event charged prompt leptons.

### CODE REVIEW

**Comment by perrozzi** [comment]
please test

File: GeneratorInterface/GenFilters/python/VBFGenJetFilter_cfi.py [hunk 1]
```diff
@@ -49,4 +49,15 @@
...
 
-)
\ No newline at end of file
+)
+
+vbfGenJetFilterD = cms.EDFilter("VBFGenJetFilter",
+
+  inputTag_GenJetCollection = cms.untracked.InputTag('ak4GenJetsNoNu'),
+
+  leadJetsNoLepMass         = cms.untracked.bool  ( True),  # Require the cut on the mass of the leading jets
```
Comment (perrozzi) [inline GeneratorInterface/GenFilters/python/VBFGenJetFilter_cfi.py]: this should probably be False

Comment (perrozzi) [reply GeneratorInterface/GenFilters/python/VBFGenJetFilter_cfi.py]: not true, should be True...
Does it work if leadJetsNoLepMass and the other parameters are not specified?

Comment (gmandorl) [reply GeneratorInterface/GenFilters/python/VBFGenJetFilter_cfi.py]: If leadJetsNoLepMass is true but no other parameters are specified the code works.
In that case, the events pass the selection whenever there are at least two jets with DR > 0.3 from all the charged hard leptons.

Comment (gmandorl) [reply GeneratorInterface/GenFilters/python/VBFGenJetFilter_cfi.py]: If leadJetsNoLepMass is not specified, it is set as false

**Comment by perrozzi** [comment]
+1

File: GeneratorInterface/GenFilters/src/VBFGenJetFilter.cc [hunk 2]
```diff
@@ -24,17 +25,41 @@ maxInvMass        (iConfig.getUntrackedParameter<double>("maxInvMass",      9999
...
 
 VBFGenJetFilter::~VBFGenJetFilter(){
   
 }
 
+
+vector<const reco::GenParticle*> VBFGenJetFilter::filterGenLeptons(const vector<reco::GenParticle>* particles){
+  vector<const reco::GenParticle*> out;
+  
+  for(unsigned i=0; i<particles->size(); i++){
```
Comment (davidlange6) [inline GeneratorInterface/GenFilters/src/VBFGenJetFilter.cc]: hi @gmandorl - please use a range based loop here (for const & p: *particles) {

File: GeneratorInterface/GenFilters/src/VBFGenJetFilter.cc [hunk 3]
```diff
@@ -24,17 +25,41 @@ maxInvMass        (iConfig.getUntrackedParameter<double>("maxInvMass",      9999
...
+      
+      const reco::GenParticle* p = &((*particles)[i]);
+      int absPdgId = abs(p->pdgId());
```
Comment (davidlange6) [inline GeneratorInterface/GenFilters/src/VBFGenJetFilter.cc]: please use std::abs here

File: GeneratorInterface/GenFilters/src/VBFGenJetFilter.cc [hunk 4]
```diff
@@ -24,17 +25,41 @@ maxInvMass        (iConfig.getUntrackedParameter<double>("maxInvMass",      9999
...
 minDeltaEta       (iConfig.getUntrackedParameter<double>("minDeltaEta",        -1.0)),
-maxDeltaEta       (iConfig.getUntrackedParameter<double>("maxDeltaEta",     99999.0))
+maxDeltaEta       (iConfig.getUntrackedParameter<double>("maxDeltaEta",     99999.0)),
+minLeadingJetsInvMass (iConfig.getUntrackedParameter<double>("minLeadingJetsInvMass",          0.0)),
+maxLeadingJetsInvMass (iConfig.getUntrackedParameter<double>("maxLeadingJetsInvMass",      99999.0)),
+deltaRJetLep          (iConfig.getUntrackedParameter<double>("deltaRJetLep",                   0.3))
 {
   
   m_inputTag_GenJetCollection = consumes<reco::GenJetCollection>(iConfig.getUntrackedParameter<edm::InputTag>("inputTag_GenJetCollection",edm::InputTag("ak5GenJetsNoNu")));
+  m_inputTag_GenParticleCollection  = consumes<reco::GenParticleCollection>(iConfig.getUntrackedParameter<edm::InputTag>("genParticles",edm::InputTag("genParticles")));
```
Comment (davidlange6) [inline GeneratorInterface/GenFilters/src/VBFGenJetFilter.cc]: this is only consumed if leadJetsNoLepMass is true - please add that conditional here.

File: GeneratorInterface/GenFilters/src/VBFGenJetFilter.cc [hunk 5]
```diff
@@ -68,6 +93,42 @@ bool VBFGenJetFilter::filter(edm::Event& iEvent, const edm::EventSetup& iSetup)
...
+      // Getting filtered generator muons
+      vector<const reco::GenParticle*> filGenLep = filterGenLeptons(genParticles);
+
+      // Getting p4 of jet with no lepton
+      vector<math::XYZTLorentzVector> genJetsWithoutLeptonsP4;
+      unsigned int jetIdx = 0;
+
+      while(genJetsWithoutLeptonsP4.size()<2 && jetIdx < filGenJets.size()) {
+          bool jetWhitoutLep = true;
+          math::XYZTLorentzVector p4J= (filGenJets[jetIdx])->p4();
```
Comment (davidlange6) [inline GeneratorInterface/GenFilters/src/VBFGenJetFilter.cc]: could this be a const &?

File: GeneratorInterface/GenFilters/src/VBFGenJetFilter.cc [hunk 6]
```diff
@@ -68,6 +93,42 @@ bool VBFGenJetFilter::filter(edm::Event& iEvent, const edm::EventSetup& iSetup)
...
+          for(unsigned int i = 0; i < filGenLep.size() && jetWhitoutLep; ++i) {
+              float dPhi = reco::deltaPhi((filGenLep[i])->p4().phi(), p4J.phi());
```
Comment (davidlange6) [inline GeneratorInterface/GenFilters/src/VBFGenJetFilter.cc]: is this code not just reco::deltaR2?

**Comment by davidlange6** [comment]
-1
pending answers to review  comments

**Comment by gmandorl** [comment]
I updated VBFGenJetFilter.cc in both the pull requests.
I also modified the Build file in order to include deltaR.h

**Comment by efeyazgan** [comment]
please test

**Comment by perrozzi** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**